redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt

```diff
@@ -126,85 +127,95 @@ internal class UIViewBox :
-            }
+      widget.modifier.forEachScoped { childModifier ->
+        when (childModifier) {
+          is HorizontalAlignment -> {
+            itemHorizontalAlignment = childModifier.alignment
+          }
 
-            is Size -> with(Density.Default) {
-              requestedWidth = childModifier.width.toPx()
-              requestedHeight = childModifier.height.toPx()
-            }
+          is VerticalAlignment -> {
+            itemVerticalAlignment = childModifier.alignment
           }
-        }
 
-        // Use requested modifiers, otherwise use the size established from sizeToFit().
-        var childWidth: CGFloat = if (requestedWidth != Double.MIN_VALUE) {
-          requestedWidth
-        } else {
-          view.frame.useContents { size.width }
-        }
-        var childHeight: CGFloat = if (requestedHeight != Double.MIN_VALUE) {
-          requestedHeight
-        } else {
-          view.frame.useContents { size.height }
-        }
+          is Width -> with(Density.Default) {
+            requestedWidth = childModifier.width.toPx()
+          }
 
-        // Compute origin and stretch if needed.
-        var x: CGFloat = 0.0
-        var y: CGFloat = 0.0
-        when (itemHorizontalAlignment) {
-          CrossAxisAlignment.Stretch -> {
-            x = 0.0
-            childWidth = frame.useContents { size.width }
+          is Height -> with(Density.Default) {
+            requestedHeight = childModifier.height.toPx()
           }
 
-          CrossAxisAlignment.Start -> x = 0.0
+          is Size -> with(Density.Default) {
+            requestedWidth = childModifier.width.toPx()
+            requestedHeight = childModifier.height.toPx()
+          }
+        }
+      }
 
-          CrossAxisAlignment.Center -> x = (frame.useContents { size.width } - childWidth) / 2.0
+      // Initialize this to the view's width before any measurement. Use the frame's dimensions
+      // if the user didn't explicitly specify one. This is the final value for Stretch alignment.
+      var viewWidth = when {
+        !requestedWidth.isNaN() -> requestedWidth
+        else -> frameWidth
+      }
+      var viewHeight = when {
+        !requestedHeight.isNaN() -> requestedHeight
+        else -> frameHeight
+      }
 
-          CrossAxisAlignment.End -> x = frame.useContents { size.width } - childWidth
-        }
-        when (itemVerticalAlignment) {
-          CrossAxisAlignment.Stretch -> {
-            y = 0.0
-            childHeight = frame.useContents { size.height }
-          }
+      // Measure the view if don't have an exact width or height.
+      val mustMeasureWidth = requestedWidth.isNaN() &&
+        itemHorizontalAlignment != CrossAxisAlignment.Stretch
+      val mustMeasureHeight = requestedHeight.isNaN() &&
+        itemVerticalAlignment != CrossAxisAlignment.Stretch
 
-          CrossAxisAlignment.Start -> y = 0.0
+      if (mustMeasureWidth || mustMeasureHeight) {
```

Comment:
This is the new part. Previously we were calling `view.sizeToFit()` at the top of this function to get a measurement as a side-effect; now we’re only doing measurement when we need to. And we’re asking for a measurement with appropriate parameters.